sdk/identity/azure-identity/azure/identity/aio/_internal/msal_credentials.py

```diff
@@ -0,0 +1,134 @@
+
+    @abc.abstractmethod
+    def get_token(self, *scopes: str) -> AccessToken:
+        pass
+
+    @abc.abstractmethod
+    def _get_app(self) -> msal.ClientApplication:
+        pass
+
+    async def _create_app(
+        self, cls: "Type[msal.ClientApplication]", executor: "Optional[Executor]" = None
+    ) -> msal.ClientApplication:
+        """Creates an MSAL application, patching msal.authority to use an azure-core pipeline during tenant discovery"""
+
+        if self._msal_app:
+            return self._msal_app
+
+        initializer = functools.partial(
+            cls,
+            client_id=self._client_id,
+            client_credential=self._client_credential,
+            authority=self._authority,
+            validate_authority=False,
+        )
+
+        loop = asyncio.get_event_loop()
+
+        # patch msal.authority so the MSAL app's init's network requests are received by the adapter
+        with mock.patch("msal.authority.requests", self._adapter) as adapter:
+            # We want the adapter to schedule transport on the event loop executing this method, so we need to pass
+            # the loop to the adapter. No argument passed to the MSAL app's init is passed through to the transport.
+            # So, we reluctantly use an instance attribute on the adapter. This should be safe because this method
+            # is called only from _get_app, and therefore within an asyncio Lock.
+            adapter.loop = loop
+            app = await loop.run_in_executor(executor, initializer)
+
+        # replace the client's requests.Session with adapter
+        app.client.session = self._adapter
+
+        return app
+
+
+class ConfidentialClientCredential(MsalCredential):
+    """Wraps an MSAL ConfidentialClientApplication with the TokenCredential API"""
+
+    async def get_token(self, *scopes: str, executor: "Optional[Executor]" = None):
+        """Get a token from the wrapped MSAL application"""
+
+        scopes = list(scopes)
+        now = int(time.time())
+
+        # try to get a cached access token or if a refresh token is cached, redeem it for an access token
+        try:
+            app = await self._get_app(executor)
+            # TODO: if we have a refresh token, acquire_token_silent will use the network
+            result = app.acquire_token_silent(scopes, account=None)
+            if not result:  # cache miss -> acquire new token
+                # Provide the loop executing this code to the adapter so the adapter can schedule transport on it.
+                # This relies on MSAL passing kwargs from its user-facing API to its transport layer.
+                loop = asyncio.get_event_loop()
+                acquire_token = functools.partial(app.acquire_token_for_client, scopes=scopes, loop=loop)
+                result = await loop.run_in_executor(executor, acquire_token)
+        except Exception as ex:
+            raise ClientAuthenticationError(message=str(ex))
```

Comment:
[`raise from`](https://docs.python.org/3/reference/simple_stmts.html#the-raise-statement)?